cpp/src/gandiva/convert_timezone_holder.h

```diff
@@ -0,0 +1,282 @@
+
+namespace gandiva {
+
+class OffsetZone {
+  std::chrono::seconds offset_;
+
+ public:
+  explicit OffsetZone(std::chrono::seconds offset) : offset_(offset) {}
+
+  template <class Duration>
+  auto to_local(arrow_vendored::date::sys_time<Duration> tp) const {
+    using std::common_type_t;
+    using std::chrono::seconds;
+    using LT = local_time<std::common_type_t<Duration, seconds>>;
+    return LT((tp + offset_).time_since_epoch());
+  }
+
+  template <class Duration>
+  auto to_sys(arrow_vendored::date::local_time<Duration> tp) const {
+    using std::chrono::seconds;
+    using ST = sys_time<std::common_type_t<Duration, seconds>>;
+    return ST((tp - offset_).time_since_epoch());
+  }
+};
+
+/// Function Holder for SQL 'CONVERT_TIMEZONE'
+class GANDIVA_EXPORT ConvertTimezoneHolder : public FunctionHolder {
+ public:
+  ~ConvertTimezoneHolder() override = default;
+
+  static Status Make(const FunctionNode& node,
+                     std::shared_ptr<ConvertTimezoneHolder>* holder);
+
+  static Status Make(const std::string& srcTz, const std::string& destTz,
+                     std::shared_ptr<ConvertTimezoneHolder>* holder);
+
+  /// Return the converted timestamp
+  int64_t convert(const int64_t src_timestamp) {
+    using std::chrono::seconds;
+
+    if (dest_timezone != NULLPTR && src_timezone == NULLPTR) {
+      return dest_timezone
+          ->to_local(src_offset_tz->to_sys<milliseconds>(
+              local_time<milliseconds>(milliseconds(src_timestamp))))
+          .time_since_epoch()
+          .count();
+    } else if (dest_timezone == NULLPTR && src_timezone != NULLPTR) {
+      return dest_offset_tz
+          ->to_local(src_timezone->to_sys<milliseconds>(
+              local_time<milliseconds>(milliseconds(src_timestamp))))
+          .time_since_epoch()
+          .count();
+    } else if (dest_timezone == NULLPTR && src_timezone == NULLPTR) {
+      return dest_offset_tz
+          ->to_local(src_offset_tz->to_sys<milliseconds>(
+              local_time<milliseconds>(milliseconds(src_timestamp))))
+          .time_since_epoch()
+          .count();
+    } else {
+      return dest_timezone
+          ->to_local(src_timezone->to_sys<milliseconds>(
+              local_time<milliseconds>(milliseconds(src_timestamp))))
+          .time_since_epoch()
+          .count();
+    }
+  }
+
+  // Tracks if the timezones given could be found in IANA Timezone DB.
+  bool ok = true;
+
+ private:
+  explicit ConvertTimezoneHolder(const std::string& srcTz, const std::string& destTz) {
+    auto srcTz_abbrv_offset = abbrv_tz.find(srcTz);
+    auto destTz_abbrv_offset = abbrv_tz.find(destTz);
+
+    try {
+      if (srcTz_abbrv_offset != abbrv_tz.end()) {
+        auto secs = convert_offset_to_seconds(srcTz_abbrv_offset->second);
+        src_offset_tz = std::make_shared<OffsetZone>(secs);
+      } else {
+        if (is_timezone_shift(srcTz) ||
+            is_timezone_offset(const_cast<std::string&>(srcTz))) {
```

Comment:
You can put that condition as an else-if and the other as else

Reply:
done